In [8]:
import pandas as pd

In [9]:
variable_respuesta = pd.read_csv("estat_lfsq_pganws_en.csv")

In [10]:
variable_respuesta

DATAFLOW        LAST UPDATE freq     unit sex  \
0        ESTAT:LFSQ_PGANWS(1.0)  15/03/24 11:00:00    Q  THS_PER   F   
1        ESTAT:LFSQ_PGANWS(1.0)  15/03/24 11:00:00    Q  THS_PER   F   
2        ESTAT:LFSQ_PGANWS(1.0)  15/03/24 11:00:00    Q  THS_PER   F   
3        ESTAT:LFSQ_PGANWS(1.0)  15/03/24 11:00:00    Q  THS_PER   F   
4        ESTAT:LFSQ_PGANWS(1.0)  15/03/24 11:00:00    Q  THS_PER   F   
...                         ...                ...  ...      ...  ..   
7563129  ESTAT:LFSQ_PGANWS(1.0)  15/03/24 11:00:00    Q  THS_PER   T   
7563130  ESTAT:LFSQ_PGANWS(1.0)  15/03/24 11:00:00    Q  THS_PER   T   
7563131  ESTAT:LFSQ_PGANWS(1.0)  15/03/24 11:00:00    Q  THS_PER   T   
7563132  ESTAT:LFSQ_PGANWS(1.0)  15/03/24 11:00:00    Q  THS_PER   T   
7563133  ESTAT:LFSQ_PGANWS(1.0)  15/03/24 11:00:00    Q  THS_PER   T   

               citizen     age wstatus geo TIME_PERIOD  OBS_VALUE OBS_FLAG  
0        EU27_2020_FOR  Y15-19     ACT  AT     1998-Q1        NaN        u  
1        EU27_2020_FOR  Y15-19     ACT  AT     1999-Q1        NaN        u  
2        EU27_2020_FOR  Y15-19     ACT  AT     2000-Q1        NaN        u  
3        EU27_2020_FOR  Y15-19     ACT  AT     2000-Q2        NaN        u  
4        EU27_2020_FOR  Y15-19     ACT  AT     2001-Q1        NaN        u  
...                ...     ...     ...  ..         ...        ...      ...  
7563129          TOTAL  Y_GE75     POP  UK     2019-Q4     5395.0      NaN  
7563130          TOTAL  Y_GE75     POP  UK     2020-Q1     5420.9      NaN  
7563131          TOTAL  Y_GE75     POP  UK     2020-Q2     5441.6      NaN  
7563132          TOTAL  Y_GE75     POP  UK     2020-Q3     5501.4      NaN  
7563133          TOTAL  Y_GE75     UNK  CY     1999-Q2       28.0      NaN  

[7563134 rows x 12 columns]

In [11]:
variable_respuesta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7563134 entries, 0 to 7563133
Data columns (total 12 columns):
 #   Column       Dtype  
---  ------       -----  
 0   DATAFLOW     object 
 1   LAST UPDATE  object 
 2   freq         object 
 3   unit         object 
 4   sex          object 
 5   citizen      object 
 6   age          object 
 7   wstatus      object 
 8   geo          object 
 9   TIME_PERIOD  object 
 10  OBS_VALUE    float64
 11  OBS_FLAG     object 
dtypes: float64(1), object(11)
memory usage: 692.4+ MB


In [12]:
variable_respuesta.TIME_PERIOD.max() # 2023, cuartil 4, VALOR MAXIMO EN TIEMPO

'2023-Q4'

In [13]:
variable_respuesta.TIME_PERIOD.min() # 1998, cuartil 1, VALOR MINIMO EN TIEMPO

'1998-Q1'

In [14]:
variable_respuesta.wstatus.value_counts()

POP     1649386
INAC    1560232
ACT     1547525
EMP     1532202
UNE     1273314
UNK         475
Name: wstatus, dtype: int64

A pesar de que los datos aparecen desagregados a varios niveles, tomaremos la tasa de actividad para ciudadanos extranjeros citizen [FOR], considerando toda la población (la suma de todas las clases de edad) y sólo la población activa wstatus [ACT] para cada país de residencia geo [todas las categorías] y para cada clase de sexo [F, M] por separado.
TENEMOS QUE METER FILTROS.  Tambien hay que cambiar la fecha a date time y agrupar por año

In [15]:
# PRUEBA GDELTDOC, EJEMPLO MUY SENCILLO!!! 
!pip install gdeltdoc


[notice] A new release of pip is available: 23.3.1 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


In [17]:
from gdeltdoc import GdeltDoc, Filters
f = Filters(
    keyword="climate change",
    start_date="2020-05-10",
    end_date="2020-05-11"
)

In [19]:
gd = GdeltDoc()
articles = gd.article_search(f)
df = pd.DataFrame(articles)

In [22]:
# vamos a visualizar este ejemplo y que tipo de dataframe nos da
df

url  \
0    https://www.sueddeutsche.de/wissen/corona-klim...   
1    https://saudigazette.com.sa/article/592941/Wor...   
2    https://laverdadnoticias.com/ecologia/Covid-19...   
3    http://www.busan.com/view/busan/view.php?code=...   
4    https://www.albayan.ae/across-the-uae/news-and...   
..                                                 ...   
245  https://www.bnnbloomberg.ca/europe-envoys-defe...   
246  https://globalnews.ca/news/6920842/ticks-lyme-...   
247  https://theconversation.com/coronavirus-shows-...   
248  https://www.infolibre.es/noticias/economia/202...   
249  https://www.diariovasco.com/sociedad/ciencia/n...   

                                            url_mobile  \
0    https://www.sueddeutsche.de/wissen/corona-klim...   
1                                                        
2    https://laverdadnoticias.com/amp/ecologia/Covi...   
3                                                        
4    https://www.albayan.ae/across-the-uae/news-and...   
..                                                 ...   
245                                                      
246  https://globalnews.ca/news/6920842/ticks-lyme-...   
247  https://theconversation.com/amp/coronavirus-sh...   
248                                                      
249  https://www.diariovasco.com/sociedad/ciencia/n...   

                                                 title          seendate  \
0    Trotzt Corona : Mehrheit für kon­se­quenten Kl...  20200510T190000Z   
1    Virus should be a wake - up call to every gove...  20200510T201500Z   
2    Covid - 19 : La disminución de contaminación e...  20200510T210000Z   
3                     생태계 교란종  등검은말벌  급속 확산 … 꿀벌 다 죽인다  20200510T170000Z   
4              كورونا نقطة تحول فاصلة في تاريخ البشرية  20200510T161500Z   
..                                                 ...               ...   
245  Europe Envoys Defend Decision to Let China Med...  20200510T004500Z   
246  More ticks and more Lyme disease in Canada fut...  20200510T110000Z   
247  Coronavirus shows housing costs leave many ins...  20200510T231500Z   
248  La receta para superar la crisis : reinventar ...  20200510T053000Z   
249  La pérdida de nieve en el Tibet favorece la ex...  20200510T091500Z   

                                           socialimage                domain  \
0    https://www.sueddeutsche.de/image/sz.1.4902342...       sueddeutsche.de   
1    https://saudigazette.com.sa/uploads/images/202...   saudigazette.com.sa   
2    https://laverdadnoticias.com/__export/15891399...  laverdadnoticias.com   
3    http://www.busan.com/nas/wcms/wcms_data/photos...             busan.com   
4    https://www.albayan.ae/polopoly_fs/1.3855034.1...            albayan.ae   
..                                                 ...                   ...   
245                                                          bnnbloomberg.ca   
246  https://shawglobalnews.files.wordpress.com/202...         globalnews.ca   
247  https://images.theconversation.com/files/33356...   theconversation.com   
248  https://d3cra5ec8gdi8w.cloudfront.net/uploads/...          infolibre.es   
249  https://static1.diariovasco.com/www/multimedia...       diariovasco.com   

    language         sourcecountry  
0     German               Germany  
1    English          Saudi Arabia  
2    Spanish                Mexico  
3     Korean           South Korea  
4     Arabic  United Arab Emirates  
..       ...                   ...  
245  English                Canada  
246  English                Canada  
247  English             Australia  
248  Spanish                 Spain  
249  Spanish                 Spain  

[250 rows x 8 columns]

In [24]:
# para analizar el tono
tone_results = gd.timeline_search("timelinetone", f)
tone_results
# significado, LOS NEGATIVOS TIENEN TONOS NEGATIVOS, LOS POSITIVOS POSITIVOS Y LOS CERCANOS A CERO NEUTROS

datetime  Average Tone
0  2020-05-10 00:00:00+00:00       -3.2793
1  2020-05-10 00:15:00+00:00       -1.6852
2  2020-05-10 00:30:00+00:00        0.3103
3  2020-05-10 00:45:00+00:00       -0.8980
4  2020-05-10 01:00:00+00:00       -1.8854
..                       ...           ...
92 2020-05-10 23:00:00+00:00       -1.8009
93 2020-05-10 23:15:00+00:00       -1.0261
94 2020-05-10 23:30:00+00:00       -1.2010
95 2020-05-10 23:45:00+00:00       -1.7848
96 2020-05-11 00:00:00+00:00        0.3741

[97 rows x 2 columns]

In [26]:
# y para la popularidad
popularity_results = gd.timeline_search("timelinevol", f)
popularity_results 


datetime  Volume Intensity
0  2020-05-10 00:00:00+00:00            0.1839
1  2020-05-10 00:15:00+00:00            0.4375
2  2020-05-10 00:30:00+00:00            0.1273
3  2020-05-10 00:45:00+00:00            0.6582
4  2020-05-10 01:00:00+00:00            0.5048
..                       ...               ...
92 2020-05-10 23:00:00+00:00            0.5805
93 2020-05-10 23:15:00+00:00            0.2661
94 2020-05-10 23:30:00+00:00            0.6459
95 2020-05-10 23:45:00+00:00            0.4957
96 2020-05-11 00:00:00+00:00            0.2811

[97 rows x 2 columns]

Popularidad:
La popularidad mide la cantidad de menciones o cobertura de un tema en los artículos.
En nuestra tabla, los valores en la columna “Volume Intensity” representan la popularidad relativa de los artículos en un intervalo de tiempo específico.
Estos valores son proporcionales dentro del conjunto de artículos monitoreados por GDELT.
Rango de Popularidad:
Los valores en la columna “Volume Intensity” pueden variar entre 0 y 1.
Un valor de 0.1839 significa que, en ese intervalo de tiempo, los artículos relacionados con el tema tenían una popularidad relativa del 18.39% en comparación con el total de artículos monitoreados.

In [28]:
# OJO, TENEMOS QUE REVISAR LOS FILTROS QUE SE NOS PIDIO, QUE TENGA MAS DE 500 PALABRAS, 
# QUE TENGAN 3 DE NUESTRS TOPICS AUN POR DEFINIR EXACTAMENTE EN LOS ARTICULOS,
# tambien surgen dudas de lo de normalizarlos luego (¿¿solo entre los articulos seleccionados??)


In [1]:
# y prueba con google trends
!pip install pytrends


[notice] A new release of pip is available: 23.3.1 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


In [4]:
# PARA BUSCAR LOS TOPICS ID
from pytrends.request import TrendReq
pytrends = TrendReq(hl='en-US', tz=360)
suggs = pytrends.suggestions("Labour Market")
print(suggs)


[{'mid': '/m/011x5vx5', 'title': 'Labour Market Regulatory Authority', 'type': 'Governing body'}, {'mid': '/g/11j775y04_', 'title': 'Labour Market Impact Assessment', 'type': 'Topic'}, {'mid': '/g/121v2c3t', 'title': 'labour market subsidy', 'type': 'Topic'}, {'mid': '/g/12bpb3230', 'title': 'International Human Resource Management', 'type': 'Book'}, {'mid': '/g/1n36gmq6p', 'title': 'labour market contribution (Denmark)', 'type': 'Topic'}]


In [5]:
suggs # topics dentro de labour market

[{'mid': '/m/011x5vx5',
  'title': 'Labour Market Regulatory Authority',
  'type': 'Governing body'},
 {'mid': '/g/11j775y04_',
  'title': 'Labour Market Impact Assessment',
  'type': 'Topic'},
 {'mid': '/g/121v2c3t', 'title': 'labour market subsidy', 'type': 'Topic'},
 {'mid': '/g/12bpb3230',
  'title': 'International Human Resource Management',
  'type': 'Book'},
 {'mid': '/g/1n36gmq6p',
  'title': 'labour market contribution (Denmark)',
  'type': 'Topic'}]

In [7]:
from pytrends.request import TrendReq
pytrends = TrendReq(hl='en-US', tz=360)
suggs = pytrends.suggestions("Credit, Loans & Personal Finance")
suggs
# NO ME SALEN LOS MISMOS QUE NOS DIERON, PORQUE SUGGESTION TE RECOMIENDA TOPICS PERO NO SON TODOS LOS TOPICS


[{'mid': '/g/12bm4cc9t',
  'title': "On My Own Two Feet: A Modern Girl's Guide to Personal Finance",
  'type': 'Book by Manisha Thakor and Sharon Kedar'},
 {'mid': '/g/11gh2w4wc1',
  'title': 'Personal Finance in Your 20s and 30s For Dummies',
  'type': 'Book by Eric Tyson'},
 {'mid': '/g/11cmck_43n',
  'title': 'Discount Finance & Personal Credit Loans Mission',
  'type': 'Topic'},
 {'mid': '/g/11f3hdstcp',
  'title': 'Suze Orman: For the Young, Fabulous & Broke',
  'type': 'Topic'},
 {'mid': '/g/11lcc9mdt7',
  'title': 'Personal Finance for Teens and Young Adults: Financial Literacy Skills To Empower Your Future, Crush Your Debt & Build Smart Money Habits That Instill Lifelong Confidence',
  'type': 'Topic'}]

In [8]:
from pytrends.request import TrendReq
pytrend = TrendReq()
topics_data = pytrend.suggestions(keyword='Labour Market') # PARA METERNOS EN LABOUR MARKET
topics_data # SIGUEN SALIENDOME OTROS

[{'mid': '/m/011x5vx5',
  'title': 'Labour Market Regulatory Authority',
  'type': 'Governing body'},
 {'mid': '/g/11j775y04_',
  'title': 'Labour Market Impact Assessment',
  'type': 'Topic'},
 {'mid': '/g/121v2c3t', 'title': 'labour market subsidy', 'type': 'Topic'},
 {'mid': '/g/12bpb3230',
  'title': 'International Human Resource Management',
  'type': 'Book'},
 {'mid': '/g/1n36gmq6p',
  'title': 'labour market contribution (Denmark)',
  'type': 'Topic'}]

In [9]:
# PROBAREMOS CON LOS TOPICS DIRECTAMENTE
from pytrends.request import TrendReq
pytrend = TrendReq()

# Obtener los topics suggestions para "Unemployment benefits"
topics_data = pytrend.suggestions(keyword='Unemployment benefits')
topics_data # Y ASI PODEMOS ACCEDER A LOS TOPICS EN INDIVIDUAL, COMENTARLO CON RANDBEE
# NO PODEMOS ACCEDER A LOS TOPICS CON LAS CATEGORIAS INICIALES (EJEMPLO, INCLUYENDO LABOUR MARKET NO NOS SALEN LOS QUE RANDBEE NOS FACILITARON,
# PERO PODEMOS BUSCARLOS DIRECTAMENTE POR SUS NOMBRES.

[{'mid': '/m/0218w7', 'title': 'Unemployment benefits', 'type': 'Topic'},
 {'mid': '/g/11j5_zdslp',
  'title': 'Unemployment benefits in New Jersey',
  'type': 'Topic'},
 {'mid': '/g/11jg4j171x',
  'title': 'Unemployment benefits in Ohio',
  'type': 'Topic'},
 {'mid': '/g/11jg4s451x',
  'title': 'Unemployment benefits in New York',
  'type': 'Topic'},
 {'mid': '/g/11jg4ldpyf',
  'title': 'Unemployment benefits in Texas',
  'type': 'Topic'}]

In [16]:
from pytrends.request import TrendReq
# Configurar TrendReq
pytrend = TrendReq()

# Especificar palabras clave y otros parámetros
keywords = ['/m/0218w7']
geo = 'US'  # Cambia según la región geográfica de interés
timeframe = 'today'  # Cambia según el período de tiempo de interés


In [17]:
# Construir payload de consulta
pytrend.build_payload(kw_list=keywords, geo=geo, timeframe=timeframe)
# Obtener datos de búsqueda
search_data = pytrend.interest_over_time()
# da error, porque son muchas solicitudes en poco tiempo



ResponseError: The request failed: Google returned a response with code 400

In [13]:
# BUSCAMOS UNA FORMA DE MANEJAR ESTE ERROR, CON BATCH Y LA BIBLIOTECA TIME, PUES ES UN ERROR RELACIONADO CON HACER DEMASIADAS CONSULTAS EN POCO TIEMPO
import time
from pytrends.request import TrendReq

pytrends = TrendReq()

keywords = ['Unemployment benefits']
batch_size = 3

for i in range(0, len(keywords), batch_size):
    batch = keywords[i:i+batch_size]
    pytrends.build_payload(kw_list=batch)
    time.sleep(5)  # Esperar 5 segundoS entre lotes

In [14]:
# y ahora, el interest
interest_over_time_df = pytrends.interest_over_time()
print(interest_over_time_df)
# sigue saliendo el mismo error, CONSULTAR

TooManyRequestsError: The request failed: Google returned a response with code 429

In [ ]:
# hecho, pruebas para conseguir los topics id y con google trend directamente

In [ ]:
# NOTA, HE HECHO UNA PRUEBA CON GOOGLE TRENDS DIRECTAMENTE, Y ES MUY FACIL E INTUITIVO, 
# INTRODUCIENDO UN TOPIC, EL PAIS O LO QUE QUERAMOS. PERO NO ES CON PYTRENDS